### ISSUE TITLE
What is a col size and ratio of index distribution?

### ISSUE THREAD

**Issue by joshivm22** [facebookresearch/dlrm_datasets#7]
https://github.com/facebookresearch/dlrm_datasets/blob/7170c1d4436267ca3963dabdf5ce57772e078e86/embedding_bag/locality_stats.txt#L6

Hi,
    Thanks a lot for this dataset and readme for getting started. Can you please explain what are the following terms and how you compute them? These are the stats available in locality_stats.txt.
1. column size
2. Ratio of index distribution at different column sizes

Thanks,

**Comment by mdschatz** [comment]
Hi and thanks for the interest,

Recall that the vectors defined in each batch of these datasets define both the embedding rows to be accessed during the embedding bag operation (indices vector) and also how they should be aggregated to form outputs (offsets vector).  These quantities effectively define the information relevant to the sparse structures of the embedding bag operation.

The "column size" refers to the frequency of a particular index arising in the indices vector.  Consider the vector `indices=[0,2,4,1,4,1]`.  This establishes that embedding vectors located at rows 0, 1, 2, and 4 are used to form the output embedding vectors and further establishes that rows 1 and 4 are used twice each; how they are combined to form the output is prescribed by the offsets vector.  For this example, we can represent the column sizes for each index using the map `{0: 1, 1: 1, 2: 2, 4: 2}` where keys are lookup indices and values are the number of occurrences (or the column size). Column sizes are computed by scanning the data and tracking how many occurrences exist for a given index.

The locality stats summarize the column size distribution of each batch of data and is represented as a histogram.  Histogram bins represent a given column size range and the height corresponds to the proportion of data falling into each bin.  Taking your referenced example, the column `(16, 32]: 0.032` states that indices with a column size between 16 and 32 account for 3.2% of the data in that batch.

Note that this distribution can change depending on whether or not duplicate indices contribute multiple times to the final tally.  Each batch report provides histograms for both methods of counting in the locality_stats.txt file.  The first histogram in each set, titled `Histogram of col sizes:`, reflects counts based on duplicates contributing multiple times while the second, titled `Ratio of index distribution at different column sizes`, reflects counts based on duplicates contributing only once.

Another way to interpret the histograms is that counts in the former histogram are weighted by the number of occurrences for each index, whereas the latter one only contributes once for each index (no matter how many times it is observed in the data).

Please let me know if this sufficiently answers your questions.

**Comment by joshivm22** [comment]
@mdschatz thanks a lot for the explanation. This clearly answers my question.

**Comment by joshivm22** [comment]
@mdschatz, I have come across another question. Why are there duplicate indices in a vector of lookup indices? Is this expected from a real dataset or is it due to some anonymization process?

**Comment by mdschatz** [comment]
@joshivm22 Great question!  Duplicate indices is a rather peculiar artifact of our production data that we too have been curious to understand but haven't gotten a clear answer to.  Essentially, there are a number of reasons why duplicate indices arise in the dataset but know that the observed duplication *is* intended to be there.

Typically, this occurs due to the application design that doesn't reduce the total data moved by applying a weight to the selected embedding rows.  The reason for this can be due to reasons such as having different APIs that prevent the weighting to be defined or due to duplication coming from the application.  One example of this could be a feature tracking the top-K types of businesses based on a selection presented to the user; if two names are provided that are classified similarly, then they could produce a duplicate entry in the table.

While it is also possible that the duplication could occur due to hash collisions coming from an anonymization or model shrinking process, this would be far rarer of an occurrence.  We haven't observed this happening too much based on the production data analysis so it's less likely the reason for your observations.

Suffice it to say, the data can be messy at times but unfortunately there isn't yet a straightforward way to discern which duplication is intended and which is spurious.  As such, we took the approach of modeling the data behavior and are following up with model designers to clean this up as much as possible.

**Comment by joshivm22** [comment]
@mdschatz thanks a lot for your answer, this clearly explains the reason for my observation. 

I have two more questions:

1. I was wondering if this dataset (of 17 mini-batches) is representative of the original dataset (from which this dataset is derived). In other words, is it correct to assume the locality stats are statistically similar to that of original dataset?

2. If no or not sure, is there any plan to publicly release the statistics of the dataset? For example, models for statistical distribution of lookup indices of original dataset for different tables. A statistical model for indices that are present together in different queries.

Thanks

**Comment by mdschatz** [comment]
Yes, the locality stats reflect the stats of the production version quite closely both across tables and for individual tables as well so the provided stats should be representative of production.

Let me know if there is anything more!

**Comment by rishucoding** [comment]
@mdschatz  @yifanliu112
I am wondering if the locality_stats are for a specific table or for all the tables together ? I believe this is important as each table would have a different memory access behavior. 

Thanks

**Comment by rishucoding** [comment]
@mdschatz @yifanliu112
For indices=[0,2,4,1,4,1], shouldn't the map be {0: 1, **1: 2**, **2: 1**, 4: 2}
instead of {0: 1, 1: 1, 2: 2, 4: 2} because index 1 repeats two times and index 2 only repeats 1 time?

**Comment by yifanliu112** [comment]
> @mdschatz @yifanliu112 I am wondering if the locality_stats are for a specific table or for all the tables together ? I believe this is important as each table would have a different memory access behavior.
> 
> Thanks

The provided locality_stats are computed with all tables combined together. You are right that each table will have a different access pattern; we are providing the results just for your reference. You can compute the locality stats for each table yourself with the logic @mdschatz outlined.

**Comment by rishucoding** [comment]
Hi @yifanliu112  - thanks for the response.

I tried to replicate the locality_stats for "fbgemm_t856_bs65536.pt" and "fbgemm_t856_bs65536_0.pt". I was able to match the #indices but my analysis is not matching with the avg # of unique columns, avg colm size, and Histogram of Colm sizes. Also, could you share what does avg colm size mean? 

Should I share my script for the analysis?  Please let me know.
